### PR TITLE
[Bugfix][Spec Decode] Fix NaN handling in rejection sampler tl.argmax

### PR DESCRIPTION
**Pull request by gabriel-peracio**
## Summary

`tl.argmax` on an all-NaN block returns an out-of-range block index (pointing into the padded region beyond `num_blocks - 1`). That index is used to load from the local-argmax tensor, producing an out-of-bounds read and an illegal memory access downstream in the rejection sampler.

Two kernels in `vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py` are affected:

- `_compute_global_target_argmax` — `tl.argmax(local_max)` at line 111, result used to index `target_local_argmax_ptr` at line 113.
- `_insert_resampled_kernel` — `tl.argmax(resampled_local_max)` at line 852, result used to index `resampled_local_argmax_ptr` at line 854.

When all values in the (padded) block dimension are NaN — reachable via NaN target logits — `tl.argmax` does not guarantee an in-range result. The loaded garbage index then propagates into subsequent gathers/stores, surfacing as an illegal memory access.

**Fix:** in both kernels, map NaN to `-inf` via `tl.where` before `tl.argmax`, and clamp the resulting block index to `num_blocks - 1` with `tl.minimum` as a defensive bound. Two small, surgical changes; no behavior change for non-NaN inputs.

## Symptom / reproduction

Observed as an illegal memory access during speculative-decoding warmup (DSpark) when serving Kimi-K3 on 8xB300 with TP=8. Confirmed the root cause with a standalone Triton repro on an RTX 4090:

```python
import torch, triton, triton.language as tl

@triton.jit
def k(out_ptr, NUM_BLOCKS: tl.constexpr, PAD: tl.constexpr):
    blocks = tl.arange(0, PAD)
    vals = tl.full((PAD,), float("nan"), tl.float32)
    vals = tl.where(blocks < NUM_BLOCKS, vals, float("-inf"))
    idx = tl.argmax(vals, axis=0)
    tl.store(out_ptr, idx)

out = torch.empty(1, dtype=torch.int32, device="cuda")
k[(1,)](out, NUM_BLOCKS=4, PAD=8)
print(out.item())  # unpatched: 8 (out of range); with NaN->-inf + clamp: in [0, 3]
```

On the all-NaN input, unpatched `tl.argmax` returns `idx == num_blocks` (out of range); with the NaN→-inf mapping plus clamp the index stays in range.

## Validation

- Standalone Triton repro (RTX 4090): patched kernel returns an in-range index on all-NaN input; unpatched does not.
- Full vLLM server on 8xB300, TP=8, Kimi-K3 with DSpark speculative decoding: server boots through warmup (previously IMA), generates coherent output, acceptance length 2.5, conc=1 throughput 110–116 tok/s (+58% vs no-spec).

## Disclosure

- AI assistance (Kimi K3) was used in developing this fix. A human engineer (Gabriel Peracio, Parasail) reviewed every changed line and validated the fix on hardware as described above.
- Checked for duplicates before opening: no open PR or issue covers `tl.argmax`/NaN out-of-bounds in the rejection sampler (`rejection_sampler_utils.py`).

### CODE REVIEW

**Comment by WoosukKwon** [comment]
cc @TheEpicDolphin for review

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py [hunk 1]
```diff
@@ -108,7 +108,10 @@ def _compute_global_target_argmax(
         mask=blocks_mask,
         other=float("-inf"),
     )
+    # See _insert_resampled_kernel: NaN breaks tl.argmax index bounds.
+    local_max = tl.where(local_max != local_max, float("-inf"), local_max)
     max_block_idx = tl.argmax(local_max, axis=0)
+    max_block_idx = tl.minimum(max_block_idx, vocab_num_blocks - 1)
```
Comment (TheEpicDolphin) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py]: I think `tl.minimum` is unnecessary here because `tl.argmax` returns the left-most index for non-NaN values during ties (if `tie_break_left=True`, which it does by default).

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py [hunk 2]
```diff
@@ -849,7 +852,18 @@ def _insert_resampled_kernel(
...
+    # of resampled_local_argmax. Map NaN to -inf and clamp the index.
+    resampled_local_max = tl.where(
+        resampled_local_max != resampled_local_max,
+        float("-inf"),
+        resampled_local_max,
+    )
     resampled_max_block_idx = tl.argmax(resampled_local_max, axis=0)
+    resampled_max_block_idx = tl.minimum(
+        resampled_max_block_idx, resample_num_blocks - 1
+    )
```
Comment (TheEpicDolphin) [inline vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py]: Same as above.

**Review by TheEpicDolphin — CHANGES_REQUESTED** [review]
Thx for the fix! Left a few comments.

**Comment by ccgibson** [comment]
## Production reproduction + fix confirmation on 8×B300 (Kimi-K3, TP8, DSpark spec-decode)

We hit this in production for a week and independently traced it to the exact site described here before finding this PR. Posting the evidence since it is the kind that is hard to collect in a test rig, and because it confirms the fix under real traffic.

### Symptom

Two crash classes on `Kimi-K3` TP8, both downstream of the same poisoned value:

- **`indexSelectSmallIndex` device assert** (`Indexing.cu:1524`), killing all 8 TP ranks. The gather is the **DSpark draft head's `[163840, 256]` bf16 `markov_head` table**, fed from `last_sampled_tokens`.
- **`Xid 13` SM warp exception, "Out Of Range Address"**, byte-identical ESR across every occurrence — a CUDA coredump named `_post_update_kernel` (`input_batch.py`), specifically the frequency/presence-penalties bin-count deref `output_bin_counts + req_state_idx*stride + token_id`, which is unbounded on `token_id`.

Both consume the sampled-token slot that `_insert_resampled_kernel` writes.

### Evidence that the garbage is this OOB read

We patched a validator into `post_update` that inspects the sampled-token tensor and logs any out-of-range entry with its raw bytes. Over multiple crashes it caught **13 poisoned slots, and all 13 sat at `col == num_sampled - 1`** — precisely the slot `_insert_resampled_kernel` writes (it stores at the old `num_sampled`, then increments). Zero hits in any other column.

The captured payloads decode as sampler floats, not token ids:

| captured value | decodes as |
|---|---|
| `0x3dac3dc23ddb3e03` | four bf16 ≈ `0.084, 0.095, 0.107, 0.128` (probabilities) |
| `0x3dcf3dcd3dee3e00` | four bf16 ≈ `0.101, 0.100, 0.116, 0.125` |
| `0x40a7000040470000` | two fp32 ≈ `5.22, 3.11` (logit magnitudes) |
| `0xbf29403dbf5dbf5b` | four bf16 ≈ `−0.66, −0.74, −0.86, −0.86` (logprob magnitudes) |
| `0x7fff7fff7fff7fff` | bf16 NaN/max — the masked-logit saturation fill |

That is the content of the buffers **adjacent to `resampled_local_argmax`**, read directly — matching this PR's mechanism exactly rather than merely being consistent with it. One capture landed ~1 s before the crash; a second showed the poison in a fully-consumed row (`col=7, ns=8`).

### Two wrong fixes, and why they failed (may save someone else the detour)

Before finding this PR we assumed the slot was *never written* and that a count was published for it. Both resulting patches failed in production, and this PR's mechanism predicts both failures:

- **reordering the count publish after the token store** — irrelevant, because the slot *is* written (with OOB garbage).
- **verifying `token_id >= 0` before publishing** — passes, because the OOB-read garbage is positive.

Sentinel-initialising the buffer to `-1` also does **not** prevent the crash: the gather truncates the int64 index to u32 before its bounds check, so `-1` becomes `0xFFFFFFFF` and still asserts. (That truncation is also a silent-corruption channel: garbage whose **low 32 bits** land under the vocab passes the assert and quietly gathers the wrong embedding row. Our validator recorded poison incidence exceeding crash incidence, so this was degrading output quality without alerting — which is why we also took #50843.)

### Fix confirmation under production traffic

We cherry-picked this PR (plus #50843) onto our frozen K3 base and rolled it one box at a time across a 3-box pool under live traffic.

The discriminating moment: the **last box still without the fix** took the class-B assert (6 Xids) while **both patched boxes absorbed the same traffic clean** — 63 min and 8 min uptime, zero restarts. We rolled the third box immediately after.

```
before   ~2–4 crashes/day pool-wide; one box was crashing at ~30-minute intervals
after    3/3 boxes clean, no Xids, no asserts, no restarts since rollout
```

Environment: 8×B300, TP8, fp8 KV, `--mamba-cache-mode align`, DSpark spec-decode (`num_speculative_tokens=7`), prefix caching on, LMCache MP KV connector. Base predates the upstream K3 merge (`d20260727`), so both hunks applied cleanly as overlays.

Happy to share the coredump or the validator patch if useful. Also relevant: #42261 looks like the same fingerprint from other reporters, and #51039 reports NaN logits on K3 after long-context prefill — a plausible source of the NaN inputs this PR guards against.

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py (hunk 1 above)
Comment (gabriel-peracio) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py]: Verified against pinned triton (3.7.1, `_argmax_combine_tie_break_left`). It is as you say, with NaN mapped to -inf, ties break to the leftmost index, a real block always wins, and the all--inf case collapses to block 0.

Dropped at both sites.

File: vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py (hunk 2 above)
Comment (gabriel-peracio) [reply vllm/v1/worker/gpu/spec_decode/rejection_sampler_utils.py]: Dropped, agreed.

**Review by TheEpicDolphin — APPROVED** [review]
Thanks, LGTM

**Comment by ZacharyZcR** [comment]
Field data on this fix's real-world severity, for anyone tracing structured-output corruption back to it:

On 4×/8× A800 80GB (sm_80) + fp8 kv-cache + DSpark speculative decoding (an sm_80 downstream tree; NaN logits rows are a known hazard on that stack), the OOB this PR fixes manifested as **35–65% corruption of tool calls** — broken tool-call structural tokens and single-token repetition runs ("0000…"×139) — while plain prose looked completely normal, because an occasionally-committed arbitrary token is invisible in free text but shreds structured output. Porting just these two `tl.where` guards dropped corruption to ~20% (the remainder was unrelated, fixed by rebasing to a post-August base); with a base that includes this fix we measure 56/56 clean structured tool calls including streaming.

If you're on a fork that branched before 2026-08-06 and serve agent/tool workloads: carry this fix. It's two branchless lines and the failure mode will not show up in perplexity or prose evals.
